mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulEditProfileDialog.kt

```diff
@@ -0,0 +1,70 @@
+package ch.epfl.sdp.mobile.state
+
+import androidx.compose.foundation.MutatePriority
+import androidx.compose.foundation.MutatorMutex
+import androidx.compose.runtime.*
+import ch.epfl.sdp.mobile.application.authentication.AuthenticatedUser
+import ch.epfl.sdp.mobile.ui.setting.EditProfileNameDialog
+import ch.epfl.sdp.mobile.ui.setting.EditProfileNameDialogState
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+/**
+ * Implementation of the [EditProfileNameDialogState] interface
+ *
+ * @param user the current [AuthenticatedUser].
+ * @param scope the coroutine scope.
+ * @param onSave the callback called after we saved the changes.
+ * @param onCancel the callback called when we click the cancel button.
+ */
+class EditProfileNameDialogStateImpl(
+    user: AuthenticatedUser,
+    private val scope: CoroutineScope,
+    onSaveAction: State<() -> Unit>,
+    onCancelAction: State<() -> Unit>
+) : EditProfileNameDialogState {
+  private var user by mutableStateOf(user)
+  private val onSaveAction by onSaveAction
+  private val onCancelAction by onCancelAction
+
+  override var name by mutableStateOf(user.name)
+
+  /**
+   * A [MutatorMutex] which enforces mutual exclusion of update profile name requests. Performing a
+   * new request (by clicking the button) will cancel the currently pending request.
+   */
+  private val mutex = MutatorMutex()
+
+  override fun onSaveClick() {
+    scope.launch {
+      user.update { name(name) }
+      mutex.mutate(MutatePriority.UserInput) { onSaveAction() }
+    }
+  }
+
+  override fun onCancelClick() {
+    onCancelAction()
+  }
+}
+
+/**
+ * A stateful implementation of [EditProfileNameDialog] which uses some composition-local values to
+ * retrieve the appropriate dependencies.
+ *
+ * @param user the current [AuthenticatedUser].
+ * @param onSave the callback called after we saved the changes.
+ * @param onCancel the callback called when we click the cancel button.
+ */
+@Composable
+fun StatefulEditProfileDialog(user: AuthenticatedUser, onSave: () -> Unit, onCancel: () -> Unit) {
```

Comment:
I think `StatefulEditProfileNameDialog` is a better name, and the file name should be updated too.